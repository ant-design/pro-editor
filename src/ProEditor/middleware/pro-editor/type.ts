import { StateCreator } from 'zustand/vanilla';
import { ConfigPublicAction } from '../../store/slices/config';
import { TakeTwo, Write } from '../types/utils';

export interface ProEditorSetStateAction {
  type: unknown;

  /**
   * 如果记录，那么历史记录的名字是什么
   */
  name?: string;
  /**
   * 是否将操作记录到历史记录中
   */
  recordHistory?: boolean;
}

/**
 * 用于给注入方法添加额外的 proEditor 实例对象
 */
export interface ProEditorMiddlewareInjectMethod<Sa extends unknown[], Sr, T, A> {
  setState<A1 extends string | ProEditorSetStateAction>(...a: [...a: TakeTwo<Sa>, action?: A1]): Sr;

  proEditor: {
    undo?: () => void;
    redo?: () => void;
    // clearStorage: () => void;
    getOptions: () => Partial<ProEditorOptions<T, A>>;
  };
}
/**
 * 提供给用户的配置项
 */
export interface ProEditorOptions<S, EditorSaveState = S> {
  /** Name of the storage (must be unique) */
  name: string;
  /**
   * Filter the persisted value.
   *
   * @params state The state's value
   */
  partialize?: (state: S) => EditorSaveState;
}

// 为  mutator 注入 'pro-editor' 类型，以支持第三个配置参数

type WithProEditor<S, A> = S extends {
  getState: () => infer T;
  setState: (...a: infer Sa) => infer Sr;
}
  ? Write<S, ProEditorMiddlewareInjectMethod<Sa, Sr, T, A>>
  : never;

declare module 'zustand/vanilla' {
  interface StoreMutators<S, A> {
    ['pro-editor']: WithProEditor<S, A>;
  }
}

// 内部方法，用于保证中间件内部的类型定义安全

export type ProEditorImpl = <T>(
  storeInitializer: StateCreator<T, [['pro-editor', unknown]], []>,
  options: ProEditorOptions<T, T>,
) => StateCreator<T, [['pro-editor', unknown]], []>;

export interface InjectInternalProEditor {
  proEditor: {
    __INTERNAL_SET_CONFIG__NOT_USE_IT: ConfigPublicAction['setConfig'];
  };
}
