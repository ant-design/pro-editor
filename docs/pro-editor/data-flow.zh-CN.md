---
title: ProEditor 数据流架构
group: 基础框架
---

# 适配多人协同模式的数据流架构

在多人协同中，我们根据数据的用途可以划分出两大类数据模型：

1. **文件数据**：唯一性数据实例（ N 个人看到的都是同一份文件数据）
2. **界面视图**：每个人默认看到自己的视图，但当用户选择跟随其他人时，视图完全 follow 对方的视图）

## ProEditor 容器数据流架构

ProEditor 的数据流包含两类：

- 容器类，有且仅有一个： `ProEditorStore` ;
- 组件类，视根据未来的扩展情况进行扩张，目前有 `ProTableStore`、`ProLayoutStore`，后续可能为会扩展 `ProFromStore`、`ProChartStore` 等；

容器类 Store 只存放 Editor 相关状态，例如：

- 开发者 / 设计师模式；
- 画布 / 代码的切换；
- 当前画布交互的激活信息；
- 当前加载的 ComponentAsset 实例；
- ...

组件类 Store 会包含具体的组件配置，例如：

- 组件配置信息；
- 组件面板切换的 tabs；
- 预览器折叠状态；
- ...

这样拆分的目的就是将关注点进行分离，进而方便未来的扩展。

## 数据流流转模式

数据流连接方式，分别从容器层和组件层进行说明。

### 容器层

容器层核心的作用是调度组件层数据与外部数据源的通信。

由于 ProEditor 本身是一个组件，因此它必须同时支持受控模式与非受控模式，这对于数据流的架构就带来了很大的挑战。

由于编辑器的信息处理较为复杂，非受控模式下，数据全部由组件自行处理是没有问题。但是一旦到了受控模式，内部状态和外部状态应该如何同步才能保证数据的一致性，并让组件消费者使用上做到较为灵活，便是一件很头疼的问题。

像 `ProTableStore` 里的 columns 数据有 20+ 的 dispatch 方法， 如果这些方法如果直接暴露给组件外部消费的开发者，外部开发者怕是要炸。

在之前的踩坑过程中，如果使用 useEffect 监听外部变化 setState 进来，内部 state 变化时 onChange 出去，非常容易造成重复渲染的死循环，这对于开发者的心智负担极重。

下面的代码在较为简单的场景下可以适用，但是在我们有 10+ 操作方法，并且需要使用 reducer 的情况下，数据流的处理就会变成一团乱麻。

```tsx | pure
import { useEffect, useState } from 'react';

const useStore = (value, onChange) => {
  const [innerValue, setValue] = useState('');

  useEffect(() => {
    setValue(value);
  }, [value]);

  const innerSetState = (value) => {
    setValue(value);

    onChange(value);
  };
};
```

因此，从最终的开发上，采用以内部数据源作为 SoT ，将外部的数据源根据情况同步进内部数据源，实现状态管理。

具体的开发策略为：

1. 首先和正常开发一样，在 Wrapper 创建一个顶层数据源，下发给子组件；
2. 将组件本身作为一个非受控组件进行开发（App），不考虑任何外部受控值；
3. 使用 `StoreUpdater` 组件作为**胶水层**隔离内部应用（`App`）与 外部数据源（`config`、`onConfigChange`、`defaultConfig`等）；
4. 根据情况在 `store` 中确定需要数据交互的方法、对象，在`internalUpdate`方法中加入`OnChange`，用`syncOutSource`方法同步外部数据源到内部；

```tsx | pure
export const Container: FC = (props) => {
  const {
    config,
    onConfigChange,
    defaultConfig,

    ...res
  } = props;

  return (
    <Wrapper>
      <App {...res} />
      <StoreUpdater config={config} defaultConfig={defaultConfig} onConfigChange={onConfigChange} />
    </Wrapper>
  );
};
```
