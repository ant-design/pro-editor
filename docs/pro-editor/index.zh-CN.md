---
title: ProBuilder 配置容器
atomId: ProBuilder
group: 组件
nav:
  title: 装配器
  order: 3
---

# ProBuilder 装配器

使用 `ProBuilder` 可以快速构建出组件的配置界面。

## 代码演示

<code src='./demos/buttonAssets.tsx' compact="true" title="Button 示例"></code>

<code src='./demos/defaultAssets.tsx' title="传入 `ComponentAssets`"></code>

<code src='./demos/empty.tsx' title="无 `ComponentAssets`"></code>

<code src="./demos/controlledPresence.tsx" title="视图状态受控" description="将 `editorAwareness` 设为受控状态，可以控制当前的视窗状态。"></code>

## API

<API id="ProEditor"></API>
