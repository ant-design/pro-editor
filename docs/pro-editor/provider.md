---
title: ProEditorProvider
group: 组件
---

# ProEditorProvider

如果：

- 希望在不是 ProEditor 组件子元素的组件中使用 `useProEditor`、useConfig、`useViewport` 等 ProEditor 钩子；
- 在页面上有多个 ProEditor 实例；
- 需要在不是 ProEditor 组件子元素的组件中访问内部状态，或使用 ProEditor 所提供的数据方法。

那么你就需要用 `ProEditorProvider` 包裹相应的应用组件。
