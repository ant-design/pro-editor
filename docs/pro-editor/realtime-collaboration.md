---
title: 实时协同模块
group: 容器能力
---

# 实时协同模块

实现了多人实时协同的基础功能的模块。

- 数据同步（ yjs-zustand 中间件）
- 协同感知（ Avatar、Cursor 组件）

## Demo 演示

<code src="./demos/realtimeCollaboration/demo.tsx"></code>

## 数据同步： Yjs 中间件

复用了 [zustand-middleware-yjs](https://github.com/joebobmiles/zustand-middleware-yjs) 中间件

后面只要为 ProEditor 容器设计合适的数据同步分层结构，ProEditor 容器就可以具有多人协作的能力。

## 协同感知

zustand-middleware-yjs 并不提供 yjs 中的 Awareness 能力。因此比如在多人协同中比较常用的用户状态、指针显示等功能是自行实现的。

## 历史记录（TODO）

历史记录在基于 YJS 实现多人协同后，是属于天生自带的功能。但这个历史记录只是大量的数据，这些数据该如何合适地对用户展示，还需要设计合理的交互方案。

## 离线缓存

离线缓存 是 Local First 理念的核心之一
