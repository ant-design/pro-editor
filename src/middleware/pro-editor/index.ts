// import {  } from '@/ProEditor/store';
import { StateCreator, StoreMutatorIdentifier } from 'zustand/vanilla';
import { ProEditorImpl, ProEditorSetStateAction } from './type';

/**
 * 提供给用户的配置项
 */
export interface ProEditorOptions<S, EditorSaveState = S> {
  /** Name of the storage (must be unique) */
  name: string;
  /**
   * Filter the persisted value.
   *
   * @params state The state's value
   */
  partialize?: (state: S) => EditorSaveState;
}

const middleware: ProEditorImpl = (storeInitializer, options) => (set, get, api) => {
  const partialize = options.partialize ?? ((s) => s);
  /**
   * 记录历史
   * @param action
   */
  const updateInProEditor = (action: ProEditorSetStateAction) => {
    const nextConfig = partialize(get());

    get().proEditor.__INTERNAL_SET_CONFIG__NOT_USE_IT(nextConfig, {
      trigger: 'proEditorMiddleware',
      ...action,
    });
  };
  /*
   * Capture the initial state so that we can initialize the pro editor store to the
   * same values as the initial values of the Zustand store.
   */
  const store = storeInitializer(
    /*
     * Create a new set function that defers to the original and then passes
     * the new state to patchSharedType.
     */
    (partial, replace, action) => {
      set(partial, replace, action);
      updateInProEditor((action as any) || {});
    },
    get,
    {
      ...api,
      // Create a new setState function as we did with set.
      setState: (partial, replace, action) => {
        api.setState(partial, replace, action);

        updateInProEditor((action as any) || {});
      },
    },
  );

  // Return the initial state to create or the next middleware.
  return {
    ...store,
    proEditor: { options: { ...options, partialize } },
  };
};

export type ProEditorMiddleware = <
  T,
  Mps extends [StoreMutatorIdentifier, unknown][] = [],
  Mcs extends [StoreMutatorIdentifier, unknown][] = [],
  U = T,
>(
  initializer: StateCreator<T, [...Mps, ['pro-editor', unknown]], Mcs>,
  options: ProEditorOptions<T, U>,
) => StateCreator<T, Mps, [['pro-editor', U], ...Mcs]>;

export const proEditorMiddleware = middleware as unknown as ProEditorMiddleware;
