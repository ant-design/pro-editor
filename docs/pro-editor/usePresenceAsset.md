---
title: usePresenceAsset
group: Hooks
order: 2
---

该方法可以获取编辑器中的资产的感知视图状态

```ts
import { usePresenceAsset } from '@ant-design/pro-editor';

const getPresenceAsset = () => {
  const [presence, setPresence] = usePresenceAsset();

  useEffect(() => {
    console.log(setPresence);
  }, [presence]);

  return null;
};
```

:::warning

注：该 hook 只能在被 `ProEditorProvider` 包裹的组件中使用，或者在 `<ProEditor />` 组件的子组件中使用。

:::
