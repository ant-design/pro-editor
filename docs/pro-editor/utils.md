---
title: C2D2C 辅助函数
group:
  title: 工具类库
  order: 99
---

# generateJSXCode

该方法用于将组件属性配置拼接生成 jsx 代码。

```jsx | pure
import { generateJSXCode } from '@ant-design/pro-editor';

const config = {
  children: '默认按钮',
  target: '_blank',
  type: 'default',
  htmlType: 'button',
  danger: false,
  icon: '',
  size: 'middle',
  loading: true,
  disabled: false,
  ghost: false,
  block: false,
};

const code = generateJSXCode('Button', config);
```

结果为:

```jsx | pure
<Button
  target="_blank"
  type="default"
  htmlType="button"
  danger={false}
  size="middle"
  loading
  disabled={false}
  ghost={false}
  block={false}
>
  默认按钮
</Button>
```
