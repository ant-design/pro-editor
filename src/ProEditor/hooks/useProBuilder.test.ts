import { ProBuilderProvider, useProBuilder } from '@ant-design/pro-editor';
import { renderHook } from '@testing-library/react';

describe('useProBuilder', () => {
  it('返回正确的实例类型', () => {
    const {
      result: { current: instance },
    } = renderHook(() => useProBuilder<{ name: string }>(), {
      wrapper: ProBuilderProvider,
    });

    expect(instance).toHaveProperty('getProps');
    expect(instance).toHaveProperty('getConfig');
    expect(instance).toHaveProperty('setConfig');
    expect(instance).toHaveProperty('exportConfig');
    expect(instance).toHaveProperty('resetConfig');

    expect(instance).toHaveProperty('getViewport');
    expect(instance).toHaveProperty('setViewport');

    expect(instance).toHaveProperty('getCanvasInteraction');
    expect(instance).toHaveProperty('setCanvasInteraction');
    expect(instance).toHaveProperty('deselectCanvas');

    expect(instance).toHaveProperty('undo');
    expect(instance).toHaveProperty('redo');
    expect(instance).toHaveProperty('undoStack');
    expect(instance).toHaveProperty('redoStack');

    expect(instance).toHaveProperty('getEditorAwareness');

    expect(instance).toHaveProperty('togglePanelExpand');

    expect(instance).not.toHaveProperty('props');
    expect(instance).not.toHaveProperty('config');
    expect(instance).not.toHaveProperty('interaction');
    expect(instance).not.toHaveProperty('editorAwareness');
    expect(instance).not.toHaveProperty('assetAwareness');
    expect(instance).not.toHaveProperty('onCanvasError');
    expect(instance).not.toHaveProperty('onEditorAwarenessChange');
    expect(instance).not.toHaveProperty('onAssetAwarenessChange');
    expect(instance).not.toHaveProperty('onConfigChange');
    expect(instance).not.toHaveProperty('onInteractionChange');
    expect(instance).not.toHaveProperty('internalSetState');
    expect(instance).not.toHaveProperty('internalUpdateCanvasInteract');
    expect(instance).not.toHaveProperty('internalUpdatePresenceEditor');
    expect(instance).not.toHaveProperty('internalUpdatePresenceAsset');
    expect(instance).not.toHaveProperty('internalUpdateConfig');
  });

  it('正确获取 config 和 props', () => {
    const {
      result: { current: instance },
    } = renderHook(() => useProBuilder<{ name: string; age: number }>(), {
      wrapper: ProBuilderProvider,
    });
    const config = { name: 'John' };
    instance.setConfig(config);
    expect(instance.getConfig()).toEqual(config);
    expect(instance.getProps()).toEqual({});
  });

  it('正确获取 presenceEditor', () => {
    const {
      result: { current: instance },
    } = renderHook(useProBuilder, { wrapper: ProBuilderProvider });
    expect(instance.getEditorAwareness()).toMatchSnapshot();
  });
});
