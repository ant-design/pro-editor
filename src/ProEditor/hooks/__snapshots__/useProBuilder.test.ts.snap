// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useProBuilder > 正确获取 presenceEditor 1`] = `
{
  "panelExpand": true,
  "panelPosition": {
    "x": 0,
    "y": 0,
  },
  "panelSize": {
    "width": 340,
  },
  "viewport": {
    "x": 0,
    "y": 0,
    "zoom": 1,
  },
}
`;
